Extended template checks

danieltre23 opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
The compiler currently doesn't really generates warning.

Issue Number: #42966 

## What is the new behavior?
This PR introduces the `extended` package which should be able to generate diagnostics after type checking (integration with the compiler should come in a future PR). The API provides an interface for `TemplateCheck`s to be authored, the package also has a function to generate the extended template diagnostics. The banana in a box check, that validates the two-way binding syntax is implemented with it's unit test.

The `extended` package is private, the only modification to the public API is the new `BANANA_IN_BOX` `ErrorCode`.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

danieltre23 committed:
refactor(compiler-cli): introduce extended package to typecheck with it's API

This commit introduces //packages/compiler-cli/src/ngtsc/typecheck/extended
as a container for a new phase of diagnostics generation. The API provides an
interface for new template checks to implement and generate template diagnostics.

Refs #42966

danieltre23 committed:
refactor(compiler-cli): introduce `getExtendedTemplateDiagnosticsForComponent` function

Add function to get extended template diagnostics generated by all the template checks.

Refs #42966

danieltre23 committed:
refactor(compiler-cli): add `BananaInBoxCheck` to the template checks

Add the implementation of a Template Check that ensures the correct
use of two-way binding syntax. Generates a warning when
'([foo])="bar"' is found instead of '[(foo)]="bar"'.

Refs #42966

danieltre23 committed:
refactor(compiler-cli): move typecheck test utils into its own package

This commit moves the test utils used in the typechecking tests into its
own package. This makes them available to be used in the tests of a
different package.

Refs #42966

danieltre23 committed:
refactor(compiler-cli): add test for `BananaInBoxCheck`

Create the test package for extended_template_diagnostics and add a unit
test for the `BananaInBoxCheck`.

Refs #42966

danieltre23 committed:
refactor(compiler-cli): move `getSourceCodeForDiagnostic` to utils

Export `getSourceCodeForDiagnostic` from `ngtsc/testing` to make it
available for other packages. This will help confirm that the source
code is correct in other tests.

Refs #42966

atscott commented on packages/compiler-cli/src/ngtsc/extended_template_diagnostics/api/api.ts:
```diff
@@ -0,0 +1,54 @@
+ * information about the diagnostics to be generated.
+ */
+export interface TemplateCheck {
+  /** Unique template check code, used for configuration and searching the error */
```
```suggestion
  /** Unique template check code, used for configuration and searching the error. */
```
super nit: end the documentation sentences with a period (here everywhere else)

atscott commented on packages/compiler-cli/src/ngtsc/extended_template_diagnostics/api/api.ts:
```diff
@@ -0,0 +1,54 @@
+  /** Unique template check code, used for configuration and searching the error */
+  code: ErrorCode;
+
+  /** Category for the diagnostic, helps skip Warning checks when needed */
```
```suggestion
  /** Category for the diagnostic, helps skip `ts.DiagnosticCategory.Warning` checks when needed. */
```

atscott commented on packages/compiler-cli/src/ngtsc/extended_template_diagnostics/src/template_checker.ts:
```diff
@@ -0,0 +1,43 @@
+export function getExtendedTemplateDiagnosticsForComponent(
+    component: ts.ClassDeclaration, templateTypeChecker: TemplateTypeChecker,
+    typeChecker: ts.TypeChecker,
+    templateChecks: TemplateCheck[] = allTemplateChecks): ts.Diagnostic[]|undefined {
```
```suggestion
    templateChecks: TemplateCheck[] = allTemplateChecks): ts.Diagnostic[]|undefined {
```

Same question about `null` vs `undefined`
// cc @alxhub

danieltre23 replied:
@dgp1130 suggested to limit the scope to just `ts.Diagnostic[]` and return an empty array in cases where no diagnostics were found.

atscott commented on packages/compiler-cli/src/ngtsc/extended_template_diagnostics/src/template_checker.ts:
```diff
@@ -0,0 +1,44 @@
+
+  const ctx = {templateTypeChecker, typeChecker, component} as TemplateContext;
+
+  templateChecks.forEach(check => {
+    check.run(ctx, template)?.forEach(diag => {
+      diagnostics.push(templateTypeChecker.makeTemplateDiagnostic(
+          component, diag.sourceSpan, check.category, check.code, diag.message));
+    });
+  });
```
```
for (const check of  templateChecks) {
  const checkDiagnostics = check.run(cts, template);
  for (const diag of checkDiagnostics) {
        diagnostics.push(templateTypeChecker.makeTemplateDiagnostic(
            component, diag.sourceSpan, check.category, check.code, diag.message));
      })});
  }

```
style nit: compiler code prefers for...of instead of forEach for better debugging experience

atscott commented on packages/compiler-cli/src/ngtsc/typecheck/test/BUILD.bazel:
```diff
@@ -26,6 +26,7 @@ ts_library(
         "//packages/compiler-cli/src/ngtsc/typecheck",
         "//packages/compiler-cli/src/ngtsc/typecheck/api",
         "//packages/compiler-cli/src/ngtsc/typecheck/diagnostics",
+        "//packages/compiler-cli/src/ngtsc/typecheck/testing",
```
nit on the commit message for `refactor(compiler-cli): move typecheck test utils into it's own package` (I can't comment on the commit message)


This commit moves the test utils used in the typechecking tests into **its**
own package**. This** makes them available to be used in the tests of a
different package.

atscott commented on packages/compiler-cli/src/ngtsc/testing/src/utils.ts:
```diff
@@ -138,3 +138,8 @@ function bindingNameEquals(node: ts.BindingName, name: string): boolean {
 }
+
+export function getSourceCodeForDiagnostic(diag: ts.Diagnostic): string {
+  const text = diag.file!.text;
+  return text.substr(diag.start!, diag.length!);
```
```suggestion
  if (diag.file === undefined || diag.start === undefined || diag.length === undefined) {
    return '';
  }
  const text = diag.file.text;
  return text.substr(diag.start, diag.length);
```
Though this is pre-existing, we should avoid null assertions where possible. This particular helper function has no knowledge of whether these are defined, even if the larger context does. Or maybe it's desirable for this function to throw an error if those are `undefined`, in which case maybe we should still have the explicit check with a `throw new Error()`

danieltre23 replied:
I think it's best to avoid the null assertion.

atscott commented on packages/compiler-cli/src/ngtsc/extended_template_diagnostics/api/api.ts:
```diff
@@ -0,0 +1,54 @@
+  category: ts.DiagnosticCategory;
+
+  /** Runs check and returns information about the diagnostics to be generated */
+  run(ctx: TemplateContext, template: TmplAstNode[]): DiagnosticInfo[]|undefined;
```
@alxhub Should we be returning `null` instead of `undefined` here to follow compiler conventions?

danieltre23 replied:
Limiting the scope to return just `DiagnosticInfo[]` and an empty array in cases where non were found.

atscott commented on packages/compiler-cli/src/ngtsc/extended_template_diagnostics/src/template_checks/banana_in_box_check.ts:
```diff
@@ -0,0 +1,72 @@
+  }
+
+  override visitElement(element: TmplAstElement): void {
+    this.visitAll(element.attributes);
+    this.visitAll(element.children);
+    this.visitAll(element.references);
+    this.visitAll(element.inputs);
+    this.visitAll(element.outputs);
```
```
    this.visitAll(element.attributes);
    this.visitAll(element.inputs);
    this.visitAll(element.outputs);
    this.visitAll(element.references);
    this.visitAll(element.children);
```
It may not matter exactly for the gather of diagnostics, but typically we visit `children` last. I've rearranged the order here to match what we do in the [language service](https://github.com/angular/angular/blob/8116895eceddc4bb2ba3ab5ac9107bce4955d048/packages/language-service/ivy/template_target.ts#L338-L355).

atscott replied:
I don't think you need to visit all of these actually. From what this check is doing, it seems like you'd only need 
```
    this.visitAll(element.outputs);
    this.visitAll(element.children);
```

atscott commented on packages/compiler-cli/src/ngtsc/extended_template_diagnostics/src/template_checks/banana_in_box_check.ts:
```diff
@@ -0,0 +1,72 @@
+    nodes.forEach(node => this.visit(node));
+  }
+
+  override visitElement(element: TmplAstElement): void {
```
I believe you'll also want to add a visitor for the templates which could also have inputs/outputs/directives

https://github.com/angular/angular/blob/master/packages/language-service/ivy/template_target.ts#L328-L335

It would be good to also add a test for this case.

danieltre23 replied:
Added a test with a directive and output in the template, and overrode the `visitTemplate` function.

danieltre23 replied:
Added a test with a directive in an `ng-template` and added inputs/outputs to the `visitTemplate` function.

atscott commented on packages/compiler-cli/src/ngtsc/extended_template_diagnostics/src/template_checks/banana_in_box_check.ts:
```diff
@@ -0,0 +1,72 @@
+   * Generate diagnostic information for the cases found.
+   */
+  override visitBoundEvent(boundEvent: TmplAstBoundEvent) {
+    const regexp = new RegExp('\[[^ ]*?\]');
```
a comment for the regex could be useful.
"Matches open brace, followed by anything but spaces, terminated by a close brace"
Also, should we match this to start/end (`'^\[[^ ]*?\]$`) so we're not matching the braces in the middle, i.e. `(abc[123]def)`. It would be good to add a test for this case.

danieltre23 replied:
Added the comment, fixed the regex and added the test case.

atscott commented on packages/compiler-cli/src/ngtsc/extended_template_diagnostics/test/template_checks_spec.ts:
```diff
@@ -0,0 +1,39 @@
+
+runInEachFileSystem(() => {
+  describe('TemplateChecks', () => {
+    it('should produce banana in a box warning', () => {
```
Adding a negative test case would be a good idea too. i.e. `should not produce banana in a box warning if written correctly`

dgp1130 reviewed (changes requested):
Great work on an initial proposal here, looking really good!

Mostly just a bunch of nitpicks, a few stylistic comments, and a couple questions about edge cases. Nothing too concerning with this PR, hope we can land it soon!

danieltre23 commented on packages/compiler-cli/src/ngtsc/extended_template_diagnostics/src/template_checker.ts:
```diff
@@ -0,0 +1,45 @@
+    typeChecker: ts.TypeChecker,
+    templateChecks: TemplateCheck[] = allTemplateChecks): ts.Diagnostic[] {
+  const template = templateTypeChecker.getTemplate(component);
+  // skip checks if component has no template
```
I guess the only case this could happen would be if the `TemplateTypeChecker` didn't have information about the component for some reason.

atscott replied:
Resolving. `getTemplate` can return `null` so it has to be handled here regardless of what the reasons are for it. I don't know if it can really happen, but the compiler was made a lot more tolerant to misconfiguration so the language service can continue to work. So it might still be possible that there is no template.

dgp1130 replied:
Reasoning for this is because the language service may have to process partially-authored files, a user could write `@Component()` but not have gotten to the `templateUrl` yet. In this scenario, there is no template and we can just ignore it.

danieltre23 commented on packages/compiler-cli/src/ngtsc/testing/src/utils.ts:
```diff
@@ -138,3 +138,11 @@ function bindingNameEquals(node: ts.BindingName, name: string): boolean {
+
+export function getSourceCodeForDiagnostic(diag: ts.Diagnostic): string {
+  if (diag.file === undefined || diag.start === undefined || diag.length === undefined) {
+    return '';
```
It depends on whether you prefer to see the test fail with an error like `Expected '' to be '[foo]='bar'''` or the error that could be thrown in this function. It might be easier to track if you see which test is failing, what do you think?

danieltre23 commented on packages/compiler-cli/src/ngtsc/extended_template_diagnostics/src/template_checks/banana_in_box_check.ts:
```diff
@@ -0,0 +1,70 @@
+}
+
+class BananaVisitor extends TmplAstRecursiveVisitor {
+  diagnosticInfo: DiagnosticInfo[] = [];
```
Not sure if this would work, `visitAll()` is called while visiting elements and templates, so it would reset the `diganosticsInfo` while visiting all. I think this could be accomplished by returning `DiagnosticInfo` from `visit*` but this would mean we'd have to override every `visit` method.

danieltre23 replied:
Added a `getDiagnostics` method which resets `diagnosticsInfo` calls `visitAll()` and returns `diagnosticsInfo`.

danieltre23 commented on packages/compiler-cli/src/ngtsc/extended_template_diagnostics/src/template_checks/banana_in_box_check.ts:
```diff
@@ -0,0 +1,70 @@
+   */
+  override visitBoundEvent(boundEvent: TmplAstBoundEvent) {
+    // Matches open bracket, followed by anything but spaces, terminated by a close bracket.
+    const regexp = new RegExp(/^\[[^ ]*?\]$/);
```
This actually doesn't accept `[ ]`, but the compiler actually accepts an output with that name, so not sure how we want to handle this case.

danieltre23 replied:
Talked offline and decided to generate a diagnostic for every output which is enclosed by brackets.

danieltre23 commented on packages/compiler-cli/src/ngtsc/extended_template_diagnostics/test/template_checks_spec.ts:
```diff
@@ -0,0 +1,71 @@
+      const {program, templateTypeChecker} = setup([{
+        fileName,
+        templates: {
+          'TestCmp': '<div (abc[123]def)="var1"> </div>',
```
Yeah, the compiler does accept this as a valid name for an output.

atscott replied:
I suggested this as a stress test for the regex expression in the lint check. I don't know that it's necessarily valid, but I still think it makes sense for us to not produce a diagnostic for this case.

dgp1130 commented on packages/compiler-cli/src/ngtsc/extended_template_diagnostics/api/api.ts:
```diff
@@ -0,0 +1,57 @@
+   * Category for the diagnostic, helps skip `ts.DiagnosticCategory.Warning`
+   * checks when needed.
+   */
+  category: ts.DiagnosticCategory;
```
Talked offline, should drop this and let the `getTemplateDiagnostics...()` hard code this value.

danieltre23 commented on packages/compiler-cli/src/ngtsc/extended_template_diagnostics/src/template_checker.ts:
```diff
@@ -0,0 +1,45 @@
+      diagnostics.push(templateTypeChecker.makeTemplateDiagnostic(
+          component, diag.sourceSpan, check.category, check.code, diag.message));
+    }
+  }
```
We need to have access to the check while generating the diagnostics, that's where we get the error code from.

dgp1130 replied:
We could also put the `.map()` immediately after `check.run()`, so it is still under the `flatMap()` and has access to `check`. This starts to make the 

```typescript
return templateChecks.flatMap((check) => {
    return check.run(ctx, template).map(({ sourceSpan, message }) => templateTypeChecker.makeTemplateDiagnostic(
        component,
        diag.sourceSpan,
        check.category,
        check.code,
        diag.message,
    ));
});
```

At that point, indenting and line wrapping might be a little awkward and the linter could be difficult here. Still up to you, just showing one possible way it could be done here.

atscott replied:
Mutating the existing array is something we do quite frequently in the compiler. In particular, the code generally prefers `for...of` over `map/flatmap`. 

// cc @alxhub

alxhub replied:
Agree with @atscott, the compiler generally prefers `for..of` to the functional style, largely due to the significantly better debug-ability.

alxhub commented on packages/compiler-cli/src/ngtsc/extended_template_diagnostics/src/template_checker.ts:
```diff
@@ -0,0 +1,45 @@
+export function getExtendedTemplateDiagnosticsForComponent(
+    component: ts.ClassDeclaration, templateTypeChecker: TemplateTypeChecker,
+    typeChecker: ts.TypeChecker,
+    templateChecks: TemplateCheck[] = allTemplateChecks): ts.Diagnostic[] {
```
You can remove the default parameter value here - in the compiler we generally prefer things to be configured explicitly.

alxhub commented on packages/compiler-cli/src/ngtsc/extended_template_diagnostics/api/api.ts:
```diff
@@ -0,0 +1,51 @@
+  code: ErrorCode;
+
+  /** Runs check and returns information about the diagnostics to be generated. */
+  run(ctx: TemplateContext, template: TmplAstNode[]): DiagnosticInfo[];
```
Why not explicitly return `ts.Diagnostic` here? We want to give `TemplateCheck`s the ability to construct rich diagnostics anyway.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/test/diagnostics_spec.ts:
```diff
@@ -12,7 +12,7 @@ import {absoluteFrom, getSourceFileOrError} from '../../file_system';
 import {OptimizeFor, TypeCheckingConfig} from '../api';
 
-import {ngForDeclaration, ngForDts, setup, TestDeclaration} from './test_utils';
+import {ngForDeclaration, ngForDts, setup, TestDeclaration} from '../testing/test_utils';
```
Like with other packages, structure `testing` to contain a `src` directory with the actual code, and an `index.ts` file (at the same level as `src`) that re-exports the public utility functions. That way the import will be

```suggestion
import {ngForDeclaration, ngForDts, setup, TestDeclaration} from '../testing';
```

dgp1130 commented on packages/compiler-cli/src/ngtsc/extended_template_diagnostics/src/template_checks/banana_in_box_check.ts:
```diff
@@ -0,0 +1,84 @@
+/**
```
Per @alxhub, let's move this into `typecheck/extended/checks/banana_in_a_box/...` so each check has its own BUILD file and independent dependencies.

dgp1130 commented on packages/compiler-cli/src/ngtsc/extended_template_diagnostics/src/template_checks/banana_in_box_check.ts:
```diff
@@ -0,0 +1,84 @@
+
+  override visitTemplate(template: TmplAstTemplate) {
+    this.visitAll(template.children);
+  }
```
Is it possible to drop all the `visit{All,Element,Template}()`?

danieltre23 committed:
refactor(compiler-cli): visit inputs and outputs for elements and templates

This commit adds to the exisiting `RecursiveVisitor` and visits inputs
and outputs for elements and templates.

Refs #42966

danieltre23 committed:
refactor(compiler-cli): make `makeTemplateDiagnostic` wrapper error
specific

This commit makes the wrapper function `makeTemplateDiagnostic` take an
ErrorCode as a type for the `ts.Diagnostic`s to be generated.

Refs #42966

danieltre23 commented on packages/compiler-cli/src/ngtsc/testing/src/utils.ts:
```diff
@@ -138,3 +138,12 @@ function bindingNameEquals(node: ts.BindingName, name: string): boolean {
+export function getSourceCodeForDiagnostic(diag: ts.Diagnostic): string {
+  if (diag.file === undefined || diag.start === undefined || diag.length === undefined) {
+    throw new Error(
+        'Unable to get source code for diagnostic. File, start and/or length could be undefined.');
```
Apparently `ts.Diagnostic` is a circular structure, so `JSON.stringify()` doesn't work on it, but let me think about the best way to print more information about the diagnostic.

danieltre23 replied:
Talked offline and decided to leave it as is.

AndrewKushnir commented on goldens/public-api/compiler-cli/error_code.md:
```diff
@@ -6,6 +6,7 @@
 
 // @public (undocumented)
 export enum ErrorCode {
+    BANANA_IN_BOX = 8101,
```
I'd propose to use more official name (i.e. "two-way binding", see https://angular.io/guide/two-way-binding) :) Also in order to make this error code more consistent with other codes, I think we can add a reason, i.e. mention that it's invalid. So my proposal is to rename the error code (and may be some names in other places):
```suggestion
    INVALID_TWO_WAY_BINDING = 8101,
```

dgp1130 replied:
I get that "banana in box" is kind of a colloquial term, but it is more specific than `INVALID_TWO_WAY_BINDING`, which could be referring to any of several possible mistakes. For example, we recently added a `SPLIT_TWO_WAY_BINDING` error, which is one kind of invalid. I guess we could update this to be more specific as a two way binding error with backwards punctuation, or calling it an output binding like `BRACKED_OUTPUT_BINDING`, though that might introduce even more confusion since users clearly intended it as a two way binding.

I feel like `BANANA_IN_BOX` is at least the best understood term for this and would probably be the least confusing for developers who encounter it.

AndrewKushnir replied:
Since this is a part of the public API, I think it'd be great to stay consistent in the terms that we use (esp. if we already use `TWO_WAY_BINDING` in other cases as you pointed out) and generally have less terms. We can consider being more explicit on what went wrong, i.e.: `INVALID_TWO_WAY_BINDING_SYNTAX`.

If the consensus is to keep `BANANA_IN_BOX`, we should probably update the error message to something like `INVALID_BANANA_IN_BOX_SYNTAX` to be consistent with other errors.

dgp1130 replied:
Agreed that the error codes tend to be phrased to refer to what is wrong. I'm ok with `INVALID_BANANA_IN_BOX` / `INVALID_BANANA_IN_BOX_SYNTAX` / `BACKWARDS_BANANA_IN_BOX` / etc.

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/testing/src/utils.ts:
```diff
@@ -138,3 +138,12 @@ function bindingNameEquals(node: ts.BindingName, name: string): boolean {
+
+export function getSourceCodeForDiagnostic(diag: ts.Diagnostic): string {
+  if (diag.file === undefined || diag.start === undefined || diag.length === undefined) {
+    throw new Error(
```
I believe this is an internal message (i.e. framework users will not see this), but I think we can still improve it a bit. Currently it's unclear what file/start/length are.
```suggestion
        throw new Error(
          'Unable to get source code for diagnostic. Provided diagnostic instance doesn't contain "file", "start" and/or "length" properties.');
```

dgp1130 replied:
We actually tried earlier, but printing the diagnostic or file would give `[object Object]` and `JSON.stringify()` fails because it s a circular structure. Probably don't need to worry too much for an assertion error like this and don't need to spend too much effort trying to make it nice.

AndrewKushnir replied:
Yeah, using `JSON.stringify()` might not be possible. My proposal is to rephrase the error string to be more explicit on what the problem is (currently it's quite unclear) without additional `JSON.stringify()` calls.

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/typecheck/extended/src/template_checker.ts:
````diff
@@ -0,0 +1,64 @@
+    diagnostics.push(...check.run(ctx, template));
+  }
+
+  // Filter out duplicated diagnostics, this is possible due to the way the compiler
+  // handles desugaring. Ex.
+  //
+  // ```
+  // <div *ngIf="true" (foo)="bar">test</div>
+  // ```
+  //
+  // Desugars into
+  //
+  // ```
+  // <div *ngIf="true" (foo)="bar">
+  //   <ng-template (foo)="bar">test</div>
+  // </div>
+  // ```
+  //
+  // In this case a duplicated diagnostic could be generated for the output `foo`.
````
This example looks incorrect, the `<ng-container>` and `<div>` should be switched. The `(foo)` listener would probably appear only on the internal node, but not twice, i.e.:
```
  // <ng-template [ngIf]="true">
  //   <div (foo)="bar">test</div>
  // </ng-template>
```
See more info at https://angular.io/guide/structural-directives#structural-directive-shorthand.

I believe there is still a valid reason why this this deduplication logic was added, but it's not obvious from the example (may be we can update it or use a different one?).

dgp1130 replied:
I believe the core issue here is that the `(foo)="bar"` output binding is duplicated, but the `<div />` and `<ng-template />` might be backwards as you mentioned. @danieltre23, can we confirm the structure you're seeing in the AST and make we're showing the duplicated binding in the right spot?

danieltre23 replied:
Yeah, the issue is that the `foo` output is duplicated, @atscott suggested I add the `AST` instead of the desugared code. I think that would be clearer and more accurate since I can get it directly from the debugger.

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/typecheck/extended/src/template_checker.ts:
```diff
@@ -0,0 +1,64 @@
+
+function areDiagnosticsEqual(first: ts.Diagnostic, second: ts.Diagnostic): boolean {
+  return first.file?.fileName === second.file?.fileName && first.start === second.start &&
+      first.length == second.length && first.code == second.code;
```
```suggestion
      first.length === second.length && first.code === second.code;
```

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/typecheck/extended/src/template_checker.ts:
```diff
@@ -0,0 +1,64 @@
+  //
+  // In this case a duplicated diagnostic could be generated for the output `foo`.
+  return diagnostics.filter((diag, index) => {
+    return diagnostics.slice(0, index).every((diag2) => !areDiagnosticsEqual(diag, diag2));
```
It looks like we'd generate a number of arrays along the way here, so I'm wondering if we can move that logic to the `for-of` loop above and add new diagnostics to the `diagnostics` array in case there are no duplicates there?

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/typecheck/extended/src/template_checks/banana_in_box/index.ts:
```diff
@@ -0,0 +1,67 @@
+
+class BananaVisitor extends TmplAstRecursiveVisitor {
+  private diagnostics: ts.Diagnostic[] = [];
+  readonly ctx: TemplateContext;
+
+  constructor(ctx: TemplateContext) {
+    super();
+    this.ctx = ctx;
+  }
```
We can probably do something like this:
```suggestion
  constructor(readonly public ctx: TemplateContext) {
    super();
  }
```

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/typecheck/extended/src/template_checks/banana_in_box/index.ts:
```diff
@@ -0,0 +1,67 @@
+                            ErrorCode.BANANA_IN_BOX>(
+          this.ctx.component, boundEvent.sourceSpan, ts.DiagnosticCategory.Warning,
+          ErrorCode.BANANA_IN_BOX,
+          `In the syntax for a two-way binding the parentheses should be inside the brackets, ex. '${
```
```suggestion
          `In the two-way binding syntax the parentheses should be inside the brackets, ex. '${
```

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/typecheck/extended/src/template_checks/banana_in_box/index.ts:
```diff
@@ -0,0 +1,67 @@
+          this.ctx.component, boundEvent.sourceSpan, ts.DiagnosticCategory.Warning,
+          ErrorCode.BANANA_IN_BOX,
+          `In the syntax for a two-way binding the parentheses should be inside the brackets, ex. '${
+              expectedBoundSyntax}' instead of '${boundSyntax}'. 
```
Curious if we need the "instead of" part. Would the diagnostics message output also contain a snippet of code where the problem was detected (with `^^^^` pointing at a specific place of a template)?

danieltre23 replied:
It would contain the code snippet, so I agree is not needed.

AndrewKushnir reviewed (commented):
@danieltre23 great work, thanks for improving template diagnostics! 👍

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/typecheck/extended/src/template_checker.ts:
````diff
@@ -0,0 +1,89 @@
+// }
+// ```
+//
+// In this case a duplicated diagnostic could be generated for the output `foo`.
````
I was thinking about these duplicate diagnostics more (why they happen). If I understand correctly, they are coming from the Template that represent inline template when structural directive is in use. In this case we have a special logic in [r3_template_transform.ts](https://github.com/angular/angular/blob/master/packages/compiler/src/render3/r3_template_transform.ts#L217-L245). To create that synthetic template node (making the current node a child of a newly created one) and copy over all inputs, outputs and some other properties. So technically in the future if we decide to introduce a check for an input, we would have the same problem of having duplicates (so it's beyond just outputs).

The deduplication logic would solve the problem (for both inputs and outputs), but it looks like we are performing extra work that we discard later (or actually do even more work to look up duplicates). I'm wondering if we can update the `BananaVisitor` logic to take that situation into account and avoid running extra checks. The most simple solution might be adding `visitTemplate` call and detect if we are in the "synthetic" ng-template node, for example:

```
  override visitTemplate(template: TmplAstTemplate) {
    const isInlineTemplateNode = template.tagName !== 'ng-template';
    if (isInlineTemplateNode) {
      // Avoid visiting attributes/inputs/outputs
      this.visitAll(template.children);
      this.visitAll(template.references);
      this.visitAll(template.variables);
    } else {
      super.visitTemplate(template);
    }
  }
```
Note: the code above is just to illustrate the idea, I have not tested it and the final implementation might be different.

dgp1130 replied:
While there is only one template check for now, we do intend to scale this up to a large number of checks in the future (and hopefully third-party checks as well). This seems like a very nuanced API for an edge case that most checks likely wouldn't try or explicitly test. I agree the extra work here is awkward and not ideal, but filtering out those duplicates at least applies that logic for all template checks rather than requiring each check to manually implement the same thing.

I also don't want to block this PR due to issues with an existing compiler edge case. We can always refactor this later, maybe when we make a common base class for template checks to extend from, it can provide this behavior in a common location.

AndrewKushnir replied:
Probably I don't have full context, but my proposal is isolated to the `BananaVisitor` class, not the entire Type checking machinery. I assume that new checks that you mention would have their own visitor classes (at least this is hinted at by this change). Also having more checks in the future (build on top of the `BananaVisitor` class as an example/pattern) would cause extra work that would be discarded later (by generating duplicates and filtering them out later), so I'd like to check if we can solve it (avoid generating duplicates in the first place). 

> This seems like a very nuanced API for an edge case that most checks likely wouldn't try or explicitly test

If my understanding is correct (@alxhub or @JoostK please correct me if I'm wrong), this is not an edge case. This situation would happen for each element with a structural directive (which is quite common, esp. with the `CommonModule` directives like *ngIf and *ngFor).

> I also don't want to block this PR due to issues with an existing compiler edge case.

I don't want to delay this PR either, so we can proceed with it (once all the necessary reviews/tests are completed) and have a TODO in the code to revisit this logic as a followup.

dgp1130 replied:
> If my understanding is correct (@alxhub or @JoostK please correct me if I'm wrong), this is not an edge case. This situation would happen for each element with a structural directive (which is quite common, esp. with the `CommonModule` directives like `*ngIf` and `*ngFor`).

I was referring to this being an edge case in the AST format, I agree that `*ngIf` and `*ngFor` are quite common in user code.

> I don't want to delay this PR either, so we can proceed with it (once all the necessary reviews/tests are completed) and have a TODO in the code to revisit this logic as a followup.

@danieltre23, how about we just leave a `TODO` here for now and clean this up in a future PR where we add the common base class that can handle this case more consistently or find a way for the AST to not have this duplication in the first place.

JoostK replied:
A quick comment from mobile: angular-eslint has this same challenge with the Angular template AST, per https://github.com/angular/angular/issues/40838. I agree with Andrew's observations but haven't looked at the code in detail to have a feeling for the best approach here.

IgorMinar reviewed (approved):
LGTM! Thanks!

Reviewed-for: public-api

jessicajaniuk reviewed (approved):
LGTM :cookie:

reviewed-for: public-api

atscott reviewed (approved):
reviewed-for: public-api

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/testing/index.ts:
```diff
@@ -35,88 +35,88 @@ export function typescriptLibDts(): TestFile {
-      }
-      declare const document: Document;
-  `
+       type Partial<T> = { [P in keyof T]?: T[P]; };
```
I think we're introducing extra whitespace here?

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/extended/src/template_checks/invalid_banana_in_box/index.ts:
```diff
@@ -0,0 +1,65 @@
+      const boundSyntax = boundEvent.sourceSpan.toString();
+      const expectedBoundSyntax = boundSyntax.replace(`(${name})`, `[(${name.slice(1, -1)})]`);
+      this.diagnostics.push(
+          this.ctx.templateTypeChecker.makeTemplateDiagnostic<ErrorCode.INVALID_BANANA_IN_BOX>(
```
We shouldn't be passing `ErrorCode.INVALID_BANANA_IN_BOX` here explicitly, it should be inferred by TypeScript from the arguments to `makeTemplateDiagnostic`.

danieltre23 replied:
Addressed in #43134

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/extended/src/template_checks/invalid_banana_in_box/index.ts:
```diff
@@ -0,0 +1,65 @@
+/**
```
This should be `extended/template_checks/invalid_banana_in_box`, not under `extended/src`. I would even suggest `extended/checks/...` since `template_` is a redundant prefix in this case.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/extended/src/template_checks/invalid_banana_in_box/index.ts:
```diff
@@ -0,0 +1,65 @@
+import {TmplAstBoundEvent, TmplAstNode, TmplAstRecursiveVisitor} from '@angular/compiler';
+import * as ts from 'typescript';
+import {ErrorCode} from '../../../../../diagnostics';
+import {TemplateCheck, TemplateContext, TemplateDiagnostic} from '../../../api/api';
```
Same here as before - we should always import from the entrypoint:

```suggestion
import {TemplateCheck, TemplateContext, TemplateDiagnostic} from '../../../api';
```

danieltre23 replied:
This is fixed in #43107

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/extended/api/api.ts:
```diff
@@ -0,0 +1,47 @@
+/**
+ * A `ts.Diangostic` with a specific error code.
+ */
+export type TemplateDiagnostic<T extends ErrorCode> = ts.Diagnostic&{code: T};
```
We already have a type named `TemplateDiagnostic` which represents a diagnostic which came from a template, so this naming is pretty confusing. The type is also wrong: `ErrorCode`s are Angular error codes, but the actual `code` we put on the `ts.Diagnostic` is a transformed version of the Angular code (to differentiate our codes from TS codes). So the actual `code` won't ever be `T`. I'm surprised TS doesn't catch this somewhere.

So what we want to represent as a type here is a `ts.Diagnostic` that is a `TemplateDiagnostic`, but that is also guaranteed to have a particular Angular-specific diagnostic code. I suggest maybe `NgTemplateDiagnostic` as a name:

```typescript
type NgTemplateDiagnostic<T extends ErrorCode> = TemplateDiagnostic & {__ngCode: T};
```

This is called a "branded type" because there is no actual `__ngCode` property. We're just declaring that this type is "marked" with a specific `ErrorCode`.

`TemplateTypeChecker.makeTemplateDiagnostic` would then need to be modified to return `NgTemplateDiagnostic`s. We should probably also move this type into the main `typecheck/api` package instead of `extended`, since it'll be returned by `TemplateTypeChecker` APIs.

danieltre23 replied:
Addressed in #43134

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/extended/api/api.ts:
```diff
@@ -0,0 +1,47 @@
+  code: T;
+
+  /** Runs check and returns information about the diagnostics to be generated. */
+  run(ctx: TemplateContext, template: TmplAstNode[]): TemplateDiagnostic<T>[];
```
It's a little strange to me that `component` is passed in `TemplateContext` but the `template` nodes themselves are a separate argument. I would suggest either moving `component` out (allowing `TemplateContext` to be reused between components) or moving `template` into the `TemplateContext` and only having one argument.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/extended/src/template_checks/invalid_banana_in_box/index.ts:
```diff
@@ -0,0 +1,65 @@
+import {TmplAstBoundEvent, TmplAstNode, TmplAstRecursiveVisitor} from '@angular/compiler';
+import * as ts from 'typescript';
+import {ErrorCode} from '../../../../../diagnostics';
+import {TemplateCheck, TemplateContext, TemplateDiagnostic} from '../../../api/api';
```
Same here:

```suggestion
import {TemplateCheck, TemplateContext, TemplateDiagnostic} from '../../../api';
```

danieltre23 replied:
Fixed in #43107
